core/utils.py

```diff
@@ -1,183 +1,77 @@
+from typing import Any
 
 
 MIN_DIRECTION = 1e-32
-EQ_EPSILON = 1e-10
+"""Numbers smaller than this won't be used as the denominator during
+division."""
+
 
+def randomly_perturb_direction(
+    direction: recourse_adapter.EmbeddedSeries, ratio: float
+) -> recourse_adapter.EmbeddedSeries:
+    """Randomly changes a vector's direction while maintaining its magnitude.
 
-def recategorize_feature(column: pd.Series, inverse_category_dict: Mapping[str, Sequence[str]]) -> pd.Series:
-    """Returns a Series where some values are remapped to others.
+    The amount of random perturbation is determined by the `ratio` argument.
+    If ratio=0.5, then random noise with magnitude 50% of the original
+    direction is added. If ratio=1, then random noise with magnitude equal to
+    the original direction is added. The direction is always rescaled to have
+    its original magnitude after adding the random noise.
 
-    Given a dictionary like {'new_val': [val1, val2]}, val1 and val2 are relabeled as new_val in the new Series.
-    
     Args:
-        column: The series of data to remap.
-        inverse_category_dict: The remapping dict formatted as described above.
-        
+        direction: The vector to perturb.
+        ratio: The amount of random noise to add as a ratio of the direction's
+            original magnitude.
+
     Returns:
-        A new series with remapped values.
+        A new vector of equal magnitude to the original but with a randomly
+        perturbed direction.
     """
-    new_column = column.copy()
-    for key, val_list in inverse_category_dict.items():
-        for val in val_list:
-            new_column = np.where(new_column == val, key, new_column)
-    return new_column
-
-
-def size_normalization(dir, poi, X):
-    return dir / X.shape[0]
-
-def cosine_similarity(x1, x2):
-    return (x1@x2) / (np.sqrt(x1@x1) * np.sqrt(x2@x2))
-
-def centroid_normalization(dir, poi, X, alpha=0.7):
-    """Normalizes direction based on the distance to the data centroid."""
-    if dir@dir == 0: # if the direction is zero or very near it, return the original direction
-        return dir
-    centroid = X.mean(axis=0)
-    diff = centroid - poi
-    centroid_dist = np.sqrt(diff@diff)
-    dir = (alpha * dir * centroid_dist) / np.sqrt(dir@dir)
-    return dir
-
-def rescale_dir(dir: recourse_adapter.EmbeddedSeries, rescale_factor: float):
-    new_dir = dir * rescale_factor
-    return new_dir
-
-def privacy_perturb_dir(dir, epsilon=0.1, delta=0.01, C=1):
-    beta = np.sqrt(2*np.log(1.25/delta))
-    stdev = (beta*C**2)/epsilon
-    return dir + np.random.normal(0, stdev, size=dir.shape)
-
-def randomly_perturb_dir(dir: recourse_adapter.EmbeddedSeries, ratio: float):
-    norm = np.linalg.norm(dir)
-    noise = np.random.normal(0, 1, len(dir))
+    norm = np.linalg.norm(direction)
+    noise = np.random.normal(0, 1, len(direction))
     noise = (noise / np.linalg.norm(noise)) * ratio * norm
```

Comment:
Do you have a check here to ensure that you're not dividing by zero?

Reply:
No! This is what caused the NaN issue earlier. It will be fixed (and unit tested) in a later PR